[receiver/hostmetrics] Deprecate processesscraper

braydonk opened the pull request:
**Description:** <Describe what has changed.>
This PR deprecates the `processesscraper` in favour of adding the functionality to the `processscraper`.

**Link to tracking Issue:**
#30895 

**Testing:** 
I wrote all testing with explanations in https://gist.github.com/braydonk/124f4242299a8cb03f44c55b056450e1

**Documentation:** <Describe the documentation added.>
I added a note to the `hostmetricsreceiver` README and generated new docs for `processscraper`.

braydonk committed:
Add system.processes.* to processscraper

This commit deprecates the processesscraper in favour of adding the
functionality to the processscraper.

braydonk committed:
Add changelog entry

braydonk committed:
Repurpose PR to not be a direct deprecation

Since this receiver is in beta, I should not have directly been making a
deprecation in this PR. Instead this PR will be focused on adding the
functionality to the `processscraper` and provides a warning suggestion
to the `processesscraper` instead of directly declaring it deprecated.

braydonk committed:
Merge branch 'main' into deprecate_processes_scraper

braydonk committed:
run make goporto

dmitryax commented on receiver/hostmetricsreceiver/internal/scraper/processesscraper/processes_scraper.go:
```diff
@@ -31,20 +31,20 @@ var metricsLength = func() int {
-// scraper for Processes Metrics
-type scraper struct {
+// Scraper for Processes Metrics
+type Scraper struct {
```
Why do we need to export any of the API here if it's not being used outside?

dmitryax commented on receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:
```diff
@@ -436,3 +448,17 @@ func (s *scraper) scrapeAndAppendSignalsPendingMetric(ctx context.Context, now p
+}
+
+func (s *scraper) scrapeAndAppendSystemProcessesCount(ctx context.Context, now pcommon.Timestamp, data aggregateProcessMetadata) {
+	if s.config.MetricsBuilderConfig.Metrics.SystemProcessesCount.Enabled {
```
This check isn't needed here because `RecordSystemProcessesCountDataPoint` already has it, but it would be nice to wrap `s.getAggregateProcessMetadata(phs)` with it so we don't throw errors for disabled metrics

braydonk replied:
There is a check in `s.getAggregateProcessMetadata` that checks first for either of the `system.process.*` metric to be enabled. Is that what you were intending, or is there another way we should guard it?

dmitryax replied:
Ok, we're good then

braydonk committed:
Remove unnecessary enabled metric guard

The guard in scrapeAndAppendSystemProcessesCreated metric was already
covered by the record data point function.

braydonk committed:
Fix CI Issues

* Add system.processes.* metrics to allowed metrics in duplicate metrics
  validation test
* Change ctx arg to _ in certain functions
* Change type name to Scraper in processesscraper fallback

braydonk committed:
Run GCI and fix cross-compile

Fixed a misname in the fallback build for aggregate process data, run
gci on the aggregate_process files.

braydonk committed:
Fix another spelling error in aggregate process fallback

braydonk committed:
I will learn to spell eventually

ChrsMark commented on receiver/hostmetricsreceiver/README.md:
```diff
@@ -191,3 +191,18 @@ Currently, the hostmetrics receiver does not set any Resource attributes on the
+    scrapers:
+      process:
+        metrics:
+          system.processes.created:
```
The changes on the metrics have already been merge in SemConv with https://github.com/open-telemetry/semantic-conventions/pull/484.

I wonder if we are planning to first deprecate the scraper and then in a follow-up PR adjust the metrics as well. Is that the plan?

braydonk replied:
The original plan discussed was to make this change first, keeping it compatible with the old metric names. This was so that it would be easier to manage during the semconv transition period.

braydonk committed:
fix windows build

dmitryax commented:
As discussed on the System Semantic Conventions SIG, we want to have a way to disable all metrics in particular group. It's needed if users don't have the `process` scraper currently enabled. Otherwise, they would have to individually disable all the `process` metrics emitted by default to make the transition.
